code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/ServiceUpdatePreListener.java

```diff
@@ -0,0 +1,89 @@
+import io.cattle.platform.core.dao.InstanceDao;
+import io.cattle.platform.core.dao.NetworkDao;
+import io.cattle.platform.core.dao.impl.InstanceDaoImpl.IpAddressToServiceIndex;
+import io.cattle.platform.core.model.Service;
+import io.cattle.platform.core.model.Subnet;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPreListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessLogic;
+import io.cattle.platform.resource.pool.ResourcePoolManager;
+import io.cattle.platform.servicediscovery.api.constants.ServiceDiscoveryConstants;
+import io.cattle.platform.servicediscovery.service.ServiceDiscoveryService;
+import io.cattle.platform.util.type.CollectionUtils;
+import io.cattle.platform.util.type.Priority;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+/*
+ * This code takes care of updating retainIp address field
+ */
+public class ServiceUpdatePreListener extends AbstractObjectProcessLogic implements ProcessPreListener,
+        Priority {
+
+    @Inject
+    ServiceDiscoveryService sdService;
+    @Inject
+    InstanceDao instanceDao;
+    @Inject
+    ResourcePoolManager poolManager;
+    @Inject
+    NetworkDao ntwkDao;
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { ServiceDiscoveryConstants.PROCESS_SERVICE_UPDATE };
+    }
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Service service = (Service) state.getResource();
+        Map<String, Object> data = state.getData();
+        Boolean oldRetainIp = false;
+        if (data.get("old") != null) {
+            Map<String, Object> old = CollectionUtils.toMap(data.get("old"));
+            if (old.get(ServiceDiscoveryConstants.FIELD_SERVICE_RETAIN_IP) != null) {
+                oldRetainIp = (Boolean) old.get(ServiceDiscoveryConstants.FIELD_SERVICE_RETAIN_IP);
+            }
+        }
+        Boolean newRetainIp = DataAccessor.fieldBoolean(service, ServiceDiscoveryConstants.FIELD_SERVICE_RETAIN_IP);
+
+        boolean update = false;
+        if (newRetainIp.booleanValue() && !oldRetainIp.booleanValue()) {
+            update = true;
+        }
+        if (!update) {
+            return null;
+        }
+
+        Subnet subnet = ntwkDao.getSubnetForObject(service, NetworkConstants.KIND_HOSTONLY);
+        if (subnet == null) {
+            return null;
+        }
+
+        // get all service managed instances and their ips/indexes
+        List<IpAddressToServiceIndex> ipToIndex = instanceDao.getIpToIndex(service, subnet);
+        for (IpAddressToServiceIndex i : ipToIndex) {
+            // release from subnet
+            poolManager.releaseResource(subnet, i.getIpAddress());
```

Comment:
@ibuildthecloud race condition prone as released resource can be allocated by some other thread. Appreciate suggestion on how to avoid it